# 01 – Data Loading & Quality Checks

## Objective
- Load raw flight data from Excel / CSV
- Inspect structure, coverage, and consistency
- Identify missing values, duplicates, and anomalies
- Produce a consolidated raw dataset for cleaning

In [1]:
import pandas as pd
import numpy as np

## Load raw data

Type: Code

Purpose: Load Excel file with multiple sheets (or CSV if exported)

Combine sheets into a single DataFrame

Quick check:
- .head()
- .shape
- .info()

Outputs:
- Table with first few rows
- Number of rows and columns
- Column types

In [2]:
file_path = "../data/raw/flights_raw.xlsx"

In [3]:
sheets = pd.read_excel(file_path, sheet_name=None)

In [4]:
type(sheets)

dict

In [5]:
sheets.keys()

dict_keys(['2020-25 OTP', '2020', '2019', '2018', '2017', '2016', '2015', '2014', '2013', '2012', '2011', '2010'])

In [6]:
df_2015 = sheets['2015']
df_2015.head(10)

Route Departing Port Arriving Port  Airline      Month  \
0    Adelaide-Brisbane       Adelaide      Brisbane  Jetstar 2015-01-01   
1  Adelaide-Gold Coast       Adelaide    Gold Coast  Jetstar 2015-01-01   
2   Adelaide-Melbourne       Adelaide     Melbourne  Jetstar 2015-01-01   
3       Adelaide-Perth       Adelaide         Perth  Jetstar 2015-01-01   
4      Adelaide-Sydney       Adelaide        Sydney  Jetstar 2015-01-01   
5    Ayers Rock-Sydney     Ayers Rock        Sydney  Jetstar 2015-01-01   
6       Ballina-Sydney        Ballina        Sydney  Jetstar 2015-01-01   
7    Brisbane-Adelaide       Brisbane      Adelaide  Jetstar 2015-01-01   
8      Brisbane-Cairns       Brisbane        Cairns  Jetstar 2015-01-01   
9      Brisbane-Darwin       Brisbane        Darwin  Jetstar 2015-01-01   

   Sectors Scheduled  Sectors Flown  Cancellations  Departures On Time  \
0                 31             31              0                  30   
1                 47             46              1                  42   
2                122            116              6                 104   
3                 43             43              0                  38   
4                150            146              4                 118   
5                 25             25              0                  24   
6                 90             90              0                  68   
7                 31             31              0                  31   
8                119            117              2                  97   
9                 30             30              0                  20   

   Arrivals On Time  Departures Delayed  Arrivals Delayed  \
0                31                   1                 0   
1                43                   4                 3   
2               102                  12                14   
3                41                   5                 2   
4               123                  28                23   
5                21                   1                 4   
6                71                  22                19   
7                31                   0                 0   
8               101                  20                16   
9                28                  10                 2   

   OnTime Departures \n(%)  OnTime Arrivals \n(%)  Cancellations \n\n(%)  
0                96.774194             100.000000               0.000000  
1                91.304348              93.478261               2.127660  
2                89.655172              87.931034               4.918033  
3                88.372093              95.348837               0.000000  
4                80.821918              84.246575               2.666667  
5                96.000000              84.000000               0.000000  
6                75.555556              78.888889               0.000000  
7               100.000000             100.000000               0.000000  
8                82.905983              86.324786               1.680672  
9                66.666667              93.333333               0.000000

In [7]:
df_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6174 entries, 0 to 6173
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Route                   6174 non-null   object        
 1   Departing Port          6174 non-null   object        
 2   Arriving Port           6174 non-null   object        
 3   Airline                 6174 non-null   object        
 4   Month                   6174 non-null   datetime64[ns]
 5   Sectors Scheduled       6174 non-null   int64         
 6   Sectors Flown           6174 non-null   int64         
 7   Cancellations           6174 non-null   int64         
 8   Departures On Time      6174 non-null   int64         
 9   Arrivals On Time        6174 non-null   int64         
 10  Departures Delayed      6174 non-null   int64         
 11  Arrivals Delayed        6174 non-null   int64         
 12  OnTime Departures 
(%)  6174 non-null   float64 

In [8]:
dfs = []

for sheet_name, df in sheets.items():
    temp = df.copy()
    temp['source_sheet'] = sheet_name
    dfs.append(temp)

flights = pd.concat(dfs, ignore_index=True)

In [9]:
flights.shape

(95166, 16)

In [10]:
flights['source_sheet'].value_counts()

source_sheet
2020-25 OTP    32029
2018            6548
2017            6470
2019            6409
2016            6342
2015            6174
2014            6045
2013            5722
2012            5236
2010            5050
2011            4944
2020            4197
Name: count, dtype: int64

In [11]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95166 entries, 0 to 95165
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Route                   95164 non-null  object        
 1   Departing Port          95161 non-null  object        
 2   Arriving Port           95161 non-null  object        
 3   Airline                 95161 non-null  object        
 4   Month                   95161 non-null  datetime64[ns]
 5   Sectors Scheduled       95161 non-null  float64       
 6   Sectors Flown           95161 non-null  float64       
 7   Cancellations           95161 non-null  float64       
 8   Departures On Time      95161 non-null  float64       
 9   Arrivals On Time        95161 non-null  float64       
 10  Departures Delayed      95161 non-null  float64       
 11  Arrivals Delayed        95161 non-null  float64       
 12  OnTime Departures 
(%)  95106 non-null  object

In [12]:
flights.head()

Route Departing Port Arriving Port  Airline      Month  \
0    Adelaide-Brisbane       Adelaide      Brisbane  Jetstar 2020-01-01   
1  Adelaide-Gold Coast       Adelaide    Gold Coast  Jetstar 2020-01-01   
2   Adelaide-Melbourne       Adelaide     Melbourne  Jetstar 2020-01-01   
3       Adelaide-Perth       Adelaide         Perth  Jetstar 2020-01-01   
4      Adelaide-Sydney       Adelaide        Sydney  Jetstar 2020-01-01   

   Sectors Scheduled  Sectors Flown  Cancellations  Departures On Time  \
0               30.0           27.0            3.0                18.0   
1               31.0           30.0            1.0                15.0   
2              119.0          106.0           13.0                61.0   
3               21.0           20.0            1.0                11.0   
4               91.0           87.0            4.0                69.0   

   Arrivals On Time  Departures Delayed  Arrivals Delayed  \
0              19.0                 9.0               8.0   
1              14.0                15.0              16.0   
2              66.0                45.0              40.0   
3              16.0                 9.0               4.0   
4              69.0                18.0              18.0   

  OnTime Departures \n(%) OnTime Arrivals \n(%)  Cancellations \n\n(%)  \
0               66.666667              70.37037              10.000000   
1                      50             46.666667               3.225806   
2                57.54717             62.264151              10.924370   
3                    55.0                    80               4.761905   
4               79.310345             79.310345               4.395604   

  source_sheet  
0  2020-25 OTP  
1  2020-25 OTP  
2  2020-25 OTP  
3  2020-25 OTP  
4  2020-25 OTP

## Check “All Airlines” rows and duplicates

Purpose:
- Identify how many rows are “All Airlines” vs per airline
- Confirm there is exactly one “All Airlines” row per route/month
- Keeping only “All Airlines” rows for route-level modeling

Outputs:
- Count table: number of “All Airlines” rows per route/month
- Any duplicates flagged

In [13]:
flights['Airline'].value_counts()

Airline
All Airlines                              23945
Virgin Australia                          20647
QantasLink                                13556
Jetstar                                   13381
Qantas                                    12339
Tigerair Australia                         3860
Virgin Australia Regional Airlines         2865
Regional Express                           2338
Rex Airlines                               1573
Virgin Australia - ATR/F100 Operations      290
Skywest                                     274
Bonza                                        35
SmartLynx Australia                          31
Hinterland                                   21
virgin Australia                              6
Name: count, dtype: int64

In [14]:
all_airlines = flights[flights['Airline'].str.strip().str.lower() == 'all airlines']

In [15]:
all_airlines_counts = (
    all_airlines
    .groupby(['Route', 'Month'])
    .size()
    .reset_index(name='n_rows')
)

In [16]:
duplicates = all_airlines_counts[all_airlines_counts['n_rows'] > 1]
duplicates

Route      Month  n_rows
59     Adelaide-Alice Springs 2020-01-01       2
60     Adelaide-Alice Springs 2020-02-01       2
61     Adelaide-Alice Springs 2020-03-01       2
62     Adelaide-Alice Springs 2020-07-01       2
63     Adelaide-Alice Springs 2020-08-01       2
...                       ...        ...     ...
22761      Wagga Wagga-Sydney 2020-08-01       2
22762      Wagga Wagga-Sydney 2020-09-01       2
22763      Wagga Wagga-Sydney 2020-10-01       2
22764      Wagga Wagga-Sydney 2020-11-01       2
22765      Wagga Wagga-Sydney 2020-12-01       2

[1120 rows x 3 columns]

In [17]:
example = flights[
    (flights['Route'] == 'Adelaide-Alice Springs') &
    (flights['Month'] == '2020-01-01') &
    (flights['Airline'].str.lower() == 'all airlines')
]

example

Route Departing Port  Arriving Port       Airline  \
395    Adelaide-Alice Springs       Adelaide  Alice Springs  All Airlines   
32424  Adelaide-Alice Springs       Adelaide  Alice Springs  All Airlines   

           Month  Sectors Scheduled  Sectors Flown  Cancellations  \
395   2020-01-01               47.0           47.0            0.0   
32424 2020-01-01               47.0           47.0            0.0   

       Departures On Time  Arrivals On Time  Departures Delayed  \
395                  40.0              41.0                 7.0   
32424                40.0              41.0                 7.0   

       Arrivals Delayed OnTime Departures \n(%) OnTime Arrivals \n(%)  \
395                 6.0               85.106383             87.234043   
32424               6.0               85.106383             87.234043   

       Cancellations \n\n(%) source_sheet  
395                      0.0  2020-25 OTP  
32424                    0.0         2020

During validation, duplicate records were identified for the same route, month, and airline.
Investigation showed these duplicates originated from overlapping Excel source sheets and contained identical operational values.
Duplicates were removed, keeping a single representative observation per route–month–airline.

In [18]:
flights_no_duplicates = flights.drop_duplicates(subset=['Route', 'Month', 'Airline'], keep='first')
flights_no_duplicates

Route Departing Port Arriving Port       Airline  \
0        Adelaide-Brisbane       Adelaide      Brisbane       Jetstar   
1      Adelaide-Gold Coast       Adelaide    Gold Coast       Jetstar   
2       Adelaide-Melbourne       Adelaide     Melbourne       Jetstar   
3           Adelaide-Perth       Adelaide         Perth       Jetstar   
4          Adelaide-Sydney       Adelaide        Sydney       Jetstar   
...                    ...            ...           ...           ...   
95161  Townsville-Brisbane     Townsville      Brisbane  All Airlines   
95162    Townsville-Cairns     Townsville        Cairns  All Airlines   
95163    Townsville-Sydney     Townsville        Sydney  All Airlines   
95164   Wagga Wagga-Sydney    Wagga Wagga        Sydney  All Airlines   
95165  All Ports-All Ports      All Ports     All Ports  All Airlines   

           Month  Sectors Scheduled  Sectors Flown  Cancellations  \
0     2020-01-01               30.0           27.0            3.0   
1     2020-01-01               31.0           30.0            1.0   
2     2020-01-01              119.0          106.0           13.0   
3     2020-01-01               21.0           20.0            1.0   
4     2020-01-01               91.0           87.0            4.0   
...          ...                ...            ...            ...   
95161 2010-12-01              339.0          336.0            3.0   
95162 2010-12-01              186.0          185.0            1.0   
95163 2010-12-01               59.0           59.0            0.0   
95164 2010-12-01              221.0          221.0            0.0   
95165 2010-12-01            45394.0        44850.0          544.0   

       Departures On Time  Arrivals On Time  Departures Delayed  \
0                    18.0              19.0                 9.0   
1                    15.0              14.0                15.0   
2                    61.0              66.0                45.0   
3                    11.0              16.0                 9.0   
4                    69.0              69.0                18.0   
...                   ...               ...                 ...   
95161               267.0             241.0                69.0   
95162               121.0             111.0                64.0   
95163                34.0              39.0                25.0   
95164               184.0             165.0                37.0   
95165             34196.0           33073.0             10654.0   

       Arrivals Delayed OnTime Departures \n(%) OnTime Arrivals \n(%)  \
0                   8.0               66.666667              70.37037   
1                  16.0                      50             46.666667   
2                  40.0                57.54717             62.264151   
3                   4.0                    55.0                    80   
4                  18.0               79.310345             79.310345   
...                 ...                     ...                   ...   
95161              95.0               79.464286              71.72619   
95162              74.0               65.405405                  60.0   
95163              20.0               57.627119             66.101695   
95164              56.0               83.257919             74.660633   
95165           11777.0               76.245262              73.74136   

       Cancellations \n\n(%) source_sheet  
0                  10.000000  2020-25 OTP  
1                   3.225806  2020-25 OTP  
2                  10.924370  2020-25 OTP  
3                   4.761905  2020-25 OTP  
4                   4.395604  2020-25 OTP  
...                      ...          ...  
95161               0.884956         2010  
95162               0.537634         2010  
95163               0.000000         2010  
95164               0.000000         2010  
95165               1.198396         2010  

[90968 rows x 16 columns]

In [19]:
before = flights.shape[0]
after = flights_no_duplicates.shape[0]

print(f"Rows before: {before}")
print(f"Rows after:  {after}")
print(f"Removed:     {before - after}")

Rows before: 95166
Rows after:  90968
Removed:     4198


In [20]:
check = (
    flights_no_duplicates[flights_no_duplicates['Airline'] == 'All Airlines']
    .groupby(['Route', 'Month'])
    .size()
    .reset_index(name='n_rows')
)

check[check['n_rows'] > 1]

Empty DataFrame
Columns: [Route, Month, n_rows]
Index: []

### Checking duplicates

In [21]:
flights_all = flights_no_duplicates[
    flights_no_duplicates['Airline'].str.strip().str.lower() == 'all airlines'
].copy()

In [22]:
flights_all

Route Departing Port  Arriving Port       Airline  \
395    Adelaide-Alice Springs       Adelaide  Alice Springs  All Airlines   
396         Adelaide-Brisbane       Adelaide       Brisbane  All Airlines   
397         Adelaide-Canberra       Adelaide       Canberra  All Airlines   
398       Adelaide-Gold Coast       Adelaide     Gold Coast  All Airlines   
399        Adelaide-Melbourne       Adelaide      Melbourne  All Airlines   
...                       ...            ...            ...           ...   
95161     Townsville-Brisbane     Townsville       Brisbane  All Airlines   
95162       Townsville-Cairns     Townsville         Cairns  All Airlines   
95163       Townsville-Sydney     Townsville         Sydney  All Airlines   
95164      Wagga Wagga-Sydney    Wagga Wagga         Sydney  All Airlines   
95165     All Ports-All Ports      All Ports      All Ports  All Airlines   

           Month  Sectors Scheduled  Sectors Flown  Cancellations  \
395   2020-01-01               47.0           47.0            0.0   
396   2020-01-01              278.0          272.0            6.0   
397   2020-01-01               62.0           60.0            2.0   
398   2020-01-01               65.0           64.0            1.0   
399   2020-01-01              711.0          673.0           38.0   
...          ...                ...            ...            ...   
95161 2010-12-01              339.0          336.0            3.0   
95162 2010-12-01              186.0          185.0            1.0   
95163 2010-12-01               59.0           59.0            0.0   
95164 2010-12-01              221.0          221.0            0.0   
95165 2010-12-01            45394.0        44850.0          544.0   

       Departures On Time  Arrivals On Time  Departures Delayed  \
395                  40.0              41.0                 7.0   
396                 231.0             235.0                41.0   
397                  53.0              54.0                 7.0   
398                  48.0              45.0                16.0   
399                 471.0             465.0               202.0   
...                   ...               ...                 ...   
95161               267.0             241.0                69.0   
95162               121.0             111.0                64.0   
95163                34.0              39.0                25.0   
95164               184.0             165.0                37.0   
95165             34196.0           33073.0             10654.0   

       Arrivals Delayed OnTime Departures \n(%) OnTime Arrivals \n(%)  \
395                 6.0               85.106383             87.234043   
396                37.0               84.926471             86.397059   
397                 6.0               88.333333                    90   
398                19.0                      75               70.3125   
399               208.0               69.985141             69.093611   
...                 ...                     ...                   ...   
95161              95.0               79.464286              71.72619   
95162              74.0               65.405405                  60.0   
95163              20.0               57.627119             66.101695   
95164              56.0               83.257919             74.660633   
95165           11777.0               76.245262              73.74136   

       Cancellations \n\n(%) source_sheet  
395                 0.000000  2020-25 OTP  
396                 2.158273  2020-25 OTP  
397                 3.225806  2020-25 OTP  
398                 1.538462  2020-25 OTP  
399                 5.344585  2020-25 OTP  
...                      ...          ...  
95161               0.884956         2010  
95162               0.537634         2010  
95163               0.000000         2010  
95164               0.000000         2010  
95165               1.198396         2010  

[22825 rows x 16 columns]

In [23]:
check = (
    flights_all
    .groupby(['Route', 'Month'])
    .size()
    .reset_index(name='n_rows')
)

check[check['n_rows'] > 1]

Empty DataFrame
Columns: [Route, Month, n_rows]
Index: []

In [24]:
flights_all['Airline'].value_counts()

Airline
All Airlines    22825
Name: count, dtype: int64

In [25]:
flights_final = flights_all[
    ~(
        (flights_all['Route'].str.strip().str.lower() == 'all ports-all ports') |
        (flights_all['Departing Port'].str.strip().str.lower() == 'all ports') |
        (flights_all['Arriving Port'].str.strip().str.lower() == 'all ports')
    )
].copy()

In [26]:
flights_final.shape

(22634, 16)

After filtering to “All Airlines” and removing national aggregates (“All Ports–All Ports”), the dataset contains 22634 route–month observations.

In [27]:
route_month_counts = (
    flights_final
    .groupby(['Route', 'Month'])
    .size()
    .reset_index(name='n_rows')
)

In [28]:
route_month_counts['n_rows'].value_counts()

n_rows
1    22634
Name: count, dtype: int64

In [29]:
route_month_counts[route_month_counts['n_rows'] > 1]

Empty DataFrame
Columns: [Route, Month, n_rows]
Index: []

In [30]:
route_month_counts.isna().sum()

Route     0
Month     0
n_rows    0
dtype: int64

### Missing values assessment

We assessed missing values across all columns in the cleaned dataset (`flights_final`).

Key findings:
- Core structural columns (`Route`, `Departing Port`, `Arriving Port`, `Airline`, `Month`) have **no missing values**.
- Target-related columns (`Cancellations`, `Departures Delayed`) also have **no missing values**.

Handling decision:
- No rows were dropped due to missing values.

This ensures full temporal and route coverage is preserved.

In [31]:
flights_final.isna().sum()

Route                      0
Departing Port             0
Arriving Port              0
Airline                    0
Month                      0
Sectors Scheduled          0
Sectors Flown              0
Cancellations              0
Departures On Time         0
Arrivals On Time           0
Departures Delayed         0
Arrivals Delayed           0
OnTime Departures \n(%)    0
OnTime Arrivals \n(%)      0
Cancellations \n\n(%)      0
source_sheet               0
dtype: int64

### Conclusion
After filtering to “All Airlines” only and removing national aggregate routes, the dataset contains one unique observation per route per month, covering January 2010 to August 2025, with no duplicate route–month combinations or missing values.

## Data types verification

In [32]:
flights_final.dtypes

Route                              object
Departing Port                     object
Arriving Port                      object
Airline                            object
Month                      datetime64[ns]
Sectors Scheduled                 float64
Sectors Flown                     float64
Cancellations                     float64
Departures On Time                float64
Arrivals On Time                  float64
Departures Delayed                float64
Arrivals Delayed                  float64
OnTime Departures \n(%)            object
OnTime Arrivals \n(%)              object
Cancellations \n\n(%)             float64
source_sheet                       object
dtype: object

### Data types verification

All columns were inspected to ensure appropriate data types.

- `Month` is correctly stored as a datetime type, enabling time-based feature extraction.
- Count-based variables (e.g. Cancellations, Departures Delayed) are numeric.
- Categorical variables (Route, Departing Port, Arriving Port, Airline) are stored as object types.

No incorrect or inconsistent data types were detected.


## Validations of values and anomalies

In [33]:
# Check for impossible or suspicious values

checks = {
    'negative_cancellations': (flights_final['Cancellations'] < 0).sum(),
    'negative_departures_delayed': (flights_final['Departures Delayed'] < 0).sum(),
    'cancellations_gt_sectors': (
        flights_final['Cancellations'] > flights_final['Sectors Scheduled']
    ).sum() if 'Sectors Scheduled' in flights_final.columns else 'N/A'
}

pd.DataFrame.from_dict(checks, orient='index', columns=['count'])


count
negative_cancellations           0
negative_departures_delayed      0
cancellations_gt_sectors         0

In [34]:
flights_final[['Cancellations', 'Departures Delayed']].describe()

Cancellations  Departures Delayed
count   22634.000000        22634.000000
mean        8.034550           55.549573
std        25.973569           74.299866
min         0.000000            0.000000
25%         0.000000           16.000000
50%         2.000000           33.000000
75%         6.000000           64.000000
max      1206.000000         1045.000000

### Validation of values and anomalies

We performed logical consistency checks on key numeric variables.

- No negative values were found in cancellations or delayed departures.
- No instances of cancellations exceeding scheduled sectors were detected.
- Summary statistics indicate plausible ranges for all operational metrics.

Extreme values observed during the 2020–2022 period correspond to known COVID-19 disruptions and are treated as real-world effects rather than data errors.


## Removing pandemic months

In [35]:
# Defining pandemic period
pandemic_start = '2020-01-25'
pandemic_end = '2023-01-12'

# Creating mask for pandemic months
is_pandemic = flights_final['Month'].between(pandemic_start, pandemic_end)

In [36]:
# Keep only months outside the pandemic period
flights_no_pandemic = flights_final[~is_pandemic].copy()

# Quick check
print(f"Original rows: {len(flights_final)}")
print(f"Rows after removing pandemic months: {len(flights_no_pandemic)}")

Original rows: 22634
Rows after removing pandemic months: 18844


In [37]:
# Confirm no pandemic months remain
flights_no_pandemic[is_pandemic]  # Should return empty

/var/folders/vg/zvlgtkv968z7n8nfz46gw5tw0000gn/T/ipykernel_89036/976561088.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  flights_no_pandemic[is_pandemic]  # Should return empty


Empty DataFrame
Columns: [Route, Departing Port, Arriving Port, Airline, Month, Sectors Scheduled, Sectors Flown, Cancellations, Departures On Time, Arrivals On Time, Departures Delayed, Arrivals Delayed, OnTime Departures 
(%), OnTime Arrivals 
(%), Cancellations 

(%), source_sheet]
Index: []

In [38]:
# Check date range
flights_no_pandemic['Month'].min(), flights_no_pandemic['Month'].max()

(Timestamp('2010-01-01 00:00:00'), Timestamp('2025-11-01 00:00:00'))

### Exclusion of COVID-19 Period

To ensure that our analysis and predictive modeling reflect typical operational conditions, we excluded the months affected by the COVID-19 pandemic. The pandemic caused extreme fluctuations in flight operations, including mass cancellations and route suspensions, which would bias model learning if retained.

Period excluded:

January 2020 through January 2023 (inclusive)

Impact on dataset:

Original dataset rows: len(flights_final)

Rows after removing pandemic period: len(flights_no_pandemic)

All remaining data spans January 2010 – December 2019 and February 2023 – August 2025.

## Save cleaned master dataset

The dataset has been filtered to:

- Include only “All Airlines” rows (route-level aggregates)  
- Exclude “All Ports–All Ports” aggregate routes  
- Remove the COVID-19 pandemic period (Jan 2020 – Jan 2023)  

This cleaned dataset is saved as `data/raw/flights_raw_combined.csv` and will be used for exploratory analysis and modeling in subsequent notebooks.


In [39]:
flights_no_pandemic.to_csv("../data/raw/flights_raw_combined.csv", index=False)